### PR TITLE
Introduce a documentation decorator and parse the TypeScript AST to find doc-decorated methods.

### PR DESCRIPTION
**Pull request by nsthorat**
This PR:
* Adds a script that parses the AST starting from 'index.ts', looking for "@doc" decorators and building a JSON representation of the API (schema in api-view.ts).
* Adds a simple mustache template which renders this JSON. This is subject to change, but is extremely low overhead so will be easy to rip out (adding vue prerender adds a lot of dependencies and complexities). The majority of the complexity of this change is in the AST parser, so this shouldn't be an issue. This lives in `website/` which is described below.
* Adds a `website/` directory - this is where we'll work on the brand new website. We can generate demos separately from the homepage / API reference (instead of homepage being a special demo).
* Currently outputs the generated HTML into `/tmp/deeplearnjs-new-website/api/index.html`. This is also subject to change as we build the new website.
* Links methods to github URLs for the method.

Example: http://nsthorat.github.io/api.html

Screenshot (lots of documentation needs to be written):
image

This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/626)

### CODE REVIEW

**Comment by dsmilkov** [comment]
:lgtm_strong: This is amazing! Great job figuring out the ts compiler API !

---

Reviewed 21 of 26 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[scripts/site/make-api.sh, line 1 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4THh6niDe6ws6h2nk9:-L4THh6oCi_U_o-lIH5g:b-i818og) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L1)):*
> ```Shell
> #!/usr/bin/env bash
> ```

let's move scripts/site/* to website/*. This way we keep anything related to website in one folder that can easily be moved to its own repo.

---

*[scripts/site/make-api.sh, line 2 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TAe4qlHMWsn2PBLx7:-L4TAe4qlHMWsn2PBLx8:bk5tig0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L2)):*
> ```Shell
> #!/usr/bin/env bash
> # Copyright 2017 Google Inc. All Rights Reserved.
> ```

s/2017/2018/

---

*[scripts/site/make-api.sh, line 23 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TB3_7tLSlj_Nnq-K5:-L4TB3_7tLSlj_Nnq-K6:b-804vut) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L23)):*
> ```Shell
> node_modules/.bin/mkdirp $OUT_PATH
> 
> ./node_modules/.bin/tsc ./scripts/site/make-api.ts --target es6 --module commonjs
> ```

How about using `ts-node` (https://github.com/TypeStrong/ts-node). It works by registering the TypeScript compiler for the .ts extension, allowing node to execute ts files. Also, in that case you can remove `make-api.sh` and move `mkdirp` to the `make-api.ts` file 

---

*[scripts/site/make-api.ts, line 22 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TBnzRERvjp28LnaTw:-L4TBnzRERvjp28LnaTx:b-48szsc) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L22)):*
> ```XML
> import * as mustache from 'mustache';
> import * as ts from 'typescript';
> import {DocHeading, DocMethod, DocMethodParam, Docs, DocSubheading} from './view-api';
> ```

I'm surprised that the linter doesn't complain about this line being longer than 80

---

*[scripts/site/make-api.ts, line 27 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TBzZ0e0AAy0kRcyFP:-L4TBzZ0e0AAy0kRcyFQ:b-2v4o7t) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L27)):*
> ```XML
> const DOCUMENTATION_DECORATOR = 'doc';
> const API_TEMPLATE_PATH = './website/api/index.html';
> const SRC_ROOT = 'src/';
> ```

remove `SRC_ROOT` and inline `src/` inside `PROGRAM_ROOT` since it's only used once

---

*[scripts/site/make-api.ts, line 37 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TCYlu2FM7jAP8phys:-L4TCYlu2FM7jAP8phyt:b-17fuuq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L37)):*
> ```XML
> if (argv.rootPath == null) {
>   throw new Error(
>       `No rootPath provided. Please provide a github root path with --rootPath.`);
> ```

Rename this flag to `repoPath`, or similar in order to differentiate it from GITHUB_ROOT, or better maybe is to remove this flag and take `cwd` as the rootPath. You kinda already do this when calling this script from `make-api.sh`.

---

*[scripts/site/make-api.ts, line 77 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TH3mQsS1pXn-O1bZ-:-L4TH3mQsS1pXn-O1bZ0:bbw1zjs) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L77)):*
> ```XML
> 
> const docs: Docs = {
>   headings: docHeadings
> ```

move this above, next to docHeadings

---

*[src/math/reduction_ops.ts, line 287 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TEUqZ9c-y__DFf2dW:-L4TEUqZ9c-y__DFf2dX:bydlmz0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/src/math/reduction_ops.ts#L287)):*
> Normalization

Maybe mean, std, moments etc should go in section "Averages". Idea is from: https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.statistics.html (see headers)

---

*[src/math/softmax.ts, line 89 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-L4TFF49b8i0vGtCxvlx:-L4TFF49b8i0vGtCxvly:b-u8bd8d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/src/math/softmax.ts#L89)):*
> ```XML
>    *     which indicates the last dimension.
>    */
>   @doc('Operations', 'Classification')
> ```

This actually gets exposed as `dl.losses.softmaxCrossEntropy` while the current doc generator shows it as `dl.softmaxCrossEntropy`. If there is a way to detect that automatically, that would be great. Otherwise we can hard-code it in decorator. Something like: `@doc('Losses', 'Classification', 'losses')` 3rd argument being a namespace that gets appended to `dl.`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/626#-:-L4THw7G80LgJqtooc-a:b-q3vu56)*

**Comment by nsthorat** [comment]
Review status: 1 of 25 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/math/reduction_ops.ts, line 287 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TEUqZ9c-y__DFf2dW:-L4XAViEEa5NQ4SR1eO2:b-88yk76) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/src/math/reduction_ops.ts#L287)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> Normalization

Maybe mean, std, moments etc should go in section "Averages". Idea is from: https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.statistics.html (see headers)
</blockquote></details>

Talked offline, mirrored TF:

https://www.tensorflow.org/api_docs/python/tf/nn/moments

---

*[src/math/softmax.ts, line 89 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TFF49b8i0vGtCxvlx:-L4XNvY20E6fp-cSGgtR:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/src/math/softmax.ts#L89)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

This actually gets exposed as `dl.losses.softmaxCrossEntropy` while the current doc generator shows it as `dl.softmaxCrossEntropy`. If there is a way to detect that automatically, that would be great. Otherwise we can hard-code it in decorator. Something like: `@doc('Losses', 'Classification', 'losses')` 3rd argument being a namespace that gets appended to `dl.`
</blockquote></details>

Done.

---

*[scripts/site/make-api.sh, line 1 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4THh6niDe6ws6h2nk9:-L4X9I80zwGN_pe0lRYl:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

let's move scripts/site/* to website/*. This way we keep anything related to website in one folder that can easily be moved to its own repo.
</blockquote></details>

Done.

---

*[scripts/site/make-api.sh, line 2 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TAe4qlHMWsn2PBLx7:-L4X9K7sTzMKvwNDRF29:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L2)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

s/2017/2018/
</blockquote></details>

Done.

---

*[scripts/site/make-api.sh, line 23 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TB3_7tLSlj_Nnq-K5:-L4XGjoCdfNh-BQf_NBT:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.sh#L23)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

How about using `ts-node` (https://github.com/TypeStrong/ts-node). It works by registering the TypeScript compiler for the .ts extension, allowing node to execute ts files. Also, in that case you can remove `make-api.sh` and move `mkdirp` to the `make-api.ts` file 
</blockquote></details>

Done.

---

*[scripts/site/make-api.ts, line 22 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TBnzRERvjp28LnaTw:-L4X9syXoYhDVsCnhaBG:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L22)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

I'm surprised that the linter doesn't complain about this line being longer than 80
</blockquote></details>

Done.

---

*[scripts/site/make-api.ts, line 27 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TBzZ0e0AAy0kRcyFP:-L4X9xm031y8S-bfmbu4:b-8nlky5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L27)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove `SRC_ROOT` and inline `src/` inside `PROGRAM_ROOT` since it's only used once
</blockquote></details>

Forgot I did this so that we could pull "src" off the path in the display. This is used now.

---

*[scripts/site/make-api.ts, line 37 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TCYlu2FM7jAP8phys:-L4XAV6HsCepofVJCuEd:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L37)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Rename this flag to `repoPath`, or similar in order to differentiate it from GITHUB_ROOT, or better maybe is to remove this flag and take `cwd` as the rootPath. You kinda already do this when calling this script from `make-api.sh`.
</blockquote></details>

Done.

---

*[scripts/site/make-api.ts, line 77 at r2](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626#-L4TH3mQsS1pXn-O1bZ-:-L4XA6WzkJ34SPWlXARM:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/6711280ec3655df1164e68e1c0d980b4742dba97/scripts/site/make-api.ts#L77)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

move this above, next to docHeadings
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/626)*
